cocotb/share/include/gpi_logging.h

```diff
@@ -37,39 +37,183 @@
 #define GPILOG_EXPORT COCOTB_IMPORT
 #endif
 
+#include <cstdarg>
+
 #ifdef __cplusplus
-# define EXTERN_C_START extern "C" {
-# define EXTERN_C_END }
-#else
-# define EXTERN_C_START
-# define EXTERN_C_END
+extern "C" {
 #endif
 
-EXTERN_C_START
-
+/** Named logging level
+ *
+ *  Only logs at one of these log level will print a level name.
+ */
 enum gpi_log_levels {
-    GPIDebug = 10,
-    GPIInfo = 20,
-    GPIWarning = 30,
-    GPIError = 40,
-    GPICritical = 50
+    GPIDebug    = 10,   ///< Prints `DEBUG`    . Verbose information, useful for debugging
+    GPIInfo     = 20,   ///< Prints `INFO`     . Information about major events in the current program
+    GPIWarning  = 30,   ///< Prints `WARN`     . Encountered a recoverable bug, or information about surprising behavior
+    GPIError    = 40,   ///< Prints `ERROR`    . An unrecoverable error
+    GPICritical = 50    ///< Prints `CRITICAL` . An unrecoverable error, to be followed by immediate simulator shutdown
 };
 
+/** Logs a message at DEBUG log level using the current log handler.
+    Automatically populates arguments using information in the called context.
+ */
```

Comment:
Adding these docstrings would make a nice self-contained PR, rather that combining it with this one.

Reply:
#2000 is half of this PR if that helps. But some changes to those files were already suggested and made here.